### PR TITLE
Stake Program

### PR DESCRIPTION
**Pull request by martinwhitman**
> ⚠️ NOTE: The AuthorizeWithStake is giving me fits still, but I think it's close; obv much cleanup/documenting to do, but I wanted to confirm this is what we're looking for

| Status  | Type  | ⚠️ Core Change | Issue |
| :---: | :---: | :---: | :--: |
| In Progress | Feature | Yes (addition) | [Link](https://github.com/bmresearch/Solnet/issues/109) |

## Problem

Implementing the Stake Program from the SDK

### CODE REVIEW

**Comment by martinwhitman** [comment]
Not sure what I did to break that unit test, if anything, but I'll figure that out

File: src/Solnet.Programs/SystemProgramData.cs
Comment (murlokito) [inline src/Solnet.Programs/SystemProgramData.cs:1]: Is this really necessary? We've been trying to minimize dependencies as much as possible

File: src/Solnet.Programs/Utilities/Serialization.cs
Comment (murlokito) [inline src/Solnet.Programs/Utilities/Serialization.cs:233]: If I am not mistaken dynamic length strings have 4 bytes which are used to define the string length, this should still remain as a u32 or not?

File: src/Solnet.Rpc/Utilities/AddressExtensions.cs
Comment (murlokito) [inline src/Solnet.Rpc/Utilities/AddressExtensions.cs:23]: Can't the helper method in the utilities be used so we don't have this kind of code duplication?

Comment (murlokito) [reply src/Solnet.Rpc/Utilities/AddressExtensions.cs:23]: The same can be said for the functions below

File: src/Solnet.Programs/StakeProgram.cs
Comment (murlokito) [inline src/Solnet.Programs/StakeProgram.cs:38]: I would create a new static class for all of these sysvars

File: src/Solnet.Programs/StakeProgram.cs
Comment (murlokito) [inline src/Solnet.Programs/StakeProgram.cs:262]: In general I have not done these ```SomeTypeArgs``` and have just passed the arguments directly to the method that encodes the data for the instruction. I would prefer if we kept this with a concise style all throughout the codebase, not having to do them for every program also makes the codebase leaner despite some of the function signatures being a little more verbose sometimes

File: src/Solnet.Programs/Models/Stake/State.cs
Comment (murlokito) [inline src/Solnet.Programs/Models/Stake/State.cs:19]: I think this will be my last comment, just regarding use of snake_case instead of PascalCase, again to maintain the code style all across the codebase.

**Review by murlokito — CHANGES_REQUESTED** [review]
This looks good, I think we need to make sure about that rust string encoding wrt the length of the string being a u32 or u64 and the encoding because I think I used ASCII because of the name service, can't recall right now.

Other than that I think this is a matter of having consistent code style with the rest of the codebase (so as to not make it even harder for new contributors) and I think we should ship this

File: src/Solnet.Programs/SystemProgramData.cs
Comment (martinwhitman) [reply src/Solnet.Programs/SystemProgramData.cs:1]: Gone!

File: src/Solnet.Programs/StakeProgram.cs
Comment (martinwhitman) [reply src/Solnet.Programs/StakeProgram.cs:38]: Moved SysVars to static class located at Solnet.Programs.SysVars

File: src/Solnet.Programs/StakeProgram.cs
Comment (martinwhitman) [reply src/Solnet.Programs/StakeProgram.cs:262]: Eliminated

File: src/Solnet.Programs/Models/Stake/State.cs
Comment (martinwhitman) [reply src/Solnet.Programs/Models/Stake/State.cs:19]: I think I got the spots where I did this.

File: src/Solnet.Rpc/Utilities/AddressExtensions.cs
Comment (martinwhitman) [reply src/Solnet.Rpc/Utilities/AddressExtensions.cs:23]: Eliminated that and the write functions below but kept createwithseed bc that doesn't exist elsewhere and that class is where other things from the pubkey.rs are implemented

Comment (martinwhitman) [reply src/Solnet.Rpc/Utilities/AddressExtensions.cs:23]: moved createwithseed to the utilities helper class to avoid circular dependency

File: src/Solnet.Programs/Utilities/Serialization.cs
Comment (martinwhitman) [reply src/Solnet.Programs/Utilities/Serialization.cs:233]: I think this might be an array vs vec issue

Comment (martinwhitman) [reply src/Solnet.Programs/Utilities/Serialization.cs:233]: Or rather, a vec vs string issue?

Comment (martinwhitman) [reply src/Solnet.Programs/Utilities/Serialization.cs:233]: https://doc.rust-lang.org/stable/std/string/struct.String.html#representation

It has to have length and capacity, so 2 x u32, so you're right that u32 is the length, and that explains why I could only get the tx to go through with a u64

Comment (martinwhitman) [reply src/Solnet.Programs/Utilities/Serialization.cs:233]: I can't get it to work without it being a u64

I am not sure but it seems like this is something like what is happening where it's looking for a u64

https://github.com/solana-labs/solana/blob/e540b1cf3c329c51de60a183529d0c39ee3d86d9/programs/bpf_loader/src/serialization.rs

**Comment by martinwhitman** [comment]
I believe I have made the requested changes

**Review by murlokito — APPROVED** [review]
I only have two comments and other than these it looks very good to me:
- there's some comments missing in the StakeProgram class
- we're trying to reorganize Solnet.Programs before actually shipping v1.0 so it's less cluttered in the future, idea is to move towards a folder per program instead of having to cram everything into the `Models` folder, in this case just add a new `Stake` folder inside `Solnet.Programs` and move everything related in there

Once this is done I'll gladly merge this :)

**Comment by martinwhitman** [comment]
I have made these changes. The TryCreateWithSeed is suddenly giving AddressWithSeedMismatch, though I am not sure what has changed to make it so. Fiddling with the EncodeRustString merely gives invalid transaction errors.

File: src/Solnet.Programs/Stake/StakeProgram.cs
Comment (tiago18c) [inline src/Solnet.Programs/Stake/StakeProgram.cs:14]: Either add the links or remove the remarks. I'm assuming this would be links for the StakeProgram rust repos?

File: src/Solnet.Programs/Stake/StakeProgramData.cs
Comment (tiago18c) [inline src/Solnet.Programs/Stake/StakeProgramData.cs:49]: Please remove the `Console.WriteLine`s in the file.

File: src/Solnet.Rpc/Utilities/AddressExtensions.cs
Comment (tiago18c) [inline src/Solnet.Rpc/Utilities/AddressExtensions.cs:97]: I think this should be moved into another class. Doesn't make sense to make this public here, but since its needed to be re-used, it makes sense to move it somewhere else. E.g. A `HashingUtilities` static class ?

File: src/Solnet.Rpc/Utilities/Utils.cs
Comment (tiago18c) [inline src/Solnet.Rpc/Utilities/Utils.cs:11]: I can't find this being used anywhere, what am I missing?

File: test/Solnet.Programs.Test/Utilities/DeserializationUtilitiesTest.cs
Comment (tiago18c) [inline test/Solnet.Programs.Test/Utilities/DeserializationUtilitiesTest.cs:274]: make it explicit `expectedLength = expected.Length + sizeof(ulong)` and there's no need for the comment.

But add a comment stating that it doesn't work for all characters due to utf-8 encoding.

File: src/Solnet.Examples/StakeExample.cs
Comment (tiago18c) [inline src/Solnet.Examples/StakeExample.cs:295]: Remove commented code. These examples need to be clean, so that someone new coming to Solnet has some examples to follow and to help understand things. Commented code won't help.

**Review by tiago18c — APPROVED** [review]
It looks good.
Consider it approved once you fix these small comments!

PS: thanks for the SysVars :)

**Review by murlokito — APPROVED** [review]
🚢🚢

**Review by tiago18c — APPROVED** [review]
LGTM ;)

**Pull request merged**
